libqtile/widget/decorations.py

```diff
@@ -0,0 +1,227 @@
+        return self.parent.width
+
+    @property
+    def drawer(self):
+        return self.parent.drawer
+
+    @property
+    def ctx(self):
+        return self.parent.drawer.ctx
+
+    def draw(self, parent):
+        self.parent = parent
+
+
+class RectDecoration(_Decoration):
+    """
+        Widget decoration that draws a rectangle behind the widget contents.
+
+        Only one colour can be set but decorations can be layered to achieve
+        multi-coloured effects.
+
+        Rectangles can be drawn as just the the outline or filled. Curved corners
+        can be obtained by setting the ``radius`` parameter.
+    """
+    defaults = [
+        ("filled", False, "Whether to fill shape"),
+        ("radius", 4, "Radius for corners (0 for square)"),
+        ("dec_colour", "#000000", "Colour for decoration"),
+        ("linewidth", 2, "Line width for decoration")
+    ]  # type: List[Tuple[str, Any, str]]
+
+    def __init__(self, **config):
+        _Decoration.__init__(self, **config)
+        self.add_defaults(RectDecoration.defaults)
+
+    def draw(self, parent):
+        _Decoration.draw(self, parent)
+
+        box_height = self.height - 2 * self.padding_y
+        box_width = self.width - 2 * self.padding_x
+
+        self.drawer.set_source_rgb(self.dec_colour)
+
+        degrees = math.pi / 180.0
+
+        self.ctx.new_sub_path()
+
+        delta = self.radius + self.linewidth / 2
+
+        self.ctx.arc(
+            self.padding_x + box_width - delta,
+            self.padding_y + delta,
+            self.radius,
+            -90 * degrees,
+            0 * degrees
+        )
+
+        self.ctx.arc(
+            self.padding_x + box_width - delta,
+            self.padding_y + box_height - delta,
+            self.radius,
+            0 * degrees,
+            90 * degrees
+        )
+
+        self.ctx.arc(
+            self.padding_x + delta,
+            self.padding_y + box_height - delta,
+            self.radius,
+            90 * degrees,
+            180 * degrees
+        )
+        self.ctx.arc(
+            self.padding_x + delta,
+            self.padding_y + delta,
+            self.radius,
+            180 * degrees,
+            270 * degrees
+        )
+
+        self.ctx.close_path()
+
+        if self.filled:
+            self.ctx.fill()
+        else:
+            self.ctx.set_line_width(self.linewidth)
+            self.ctx.stroke()
+
+
+class BorderDecoration(_Decoration):
+    """
+        Widget decoration that draws a straight line on the widget border.
+        Padding can be used to adjust the position of the border further.
+
+        Only one colour can be set but decorations can be layered to achieve
+        multi-coloured effects.
+    """
+    defaults = [
+        ("colour", "#000000", "Border colour"),
+        (
+            "border_width",
+            2,
+            "Border width. Single number is all edges, or [top/bottom, left/right] or [top, bottom, left, right]"
```

Comment:
Upon first read I got a bit confused. Perhaps to be consistent with other things we can specify in all four sides (like Bar margins) we should stick to either a single number or all 4? Otherwise people might try doing the `[top/bottom, left/right]` version where it isn't supported. Along the same lines, can this be changed to `[N, E, S, W]`, to stay consistent with similar options around the codebase?

Reply:
Yeah. I did a bad job of describing this. Will adjust as you suggest.